File: core/bech32.zig
const std = @import("std");

// ─── Bech32 / Bech32m Encoder & Decoder (BIP-173 + BIP-350) ─────────────────
//
// HRP pentru OmniBus: "ob"
// Witness version 0  → Bech32  (ob1q...)  — P2WPKH / P2WSH
// Witness version 1+ → Bech32m (ob1p...)  — Taproot / viitoare versiuni
//
// Format: <hrp> "1" <data characters> <6 char checksum>
//   - data[0] = witness version (0..16)
//   - data[1..] = witness program (hash160 sau hash256) convertit în 5-bit groups

pub const Encoding = enum {
    bech32,
    bech32m,
};

/// Constantele checksum conform BIP-173 / BIP-350
const BECH32_CONST: u32 = 1;
const BECH32M_CONST: u32 = 0x2bc830a3;

const CHARSET = "qpzry9x8gf2tvdw0s3jn54khce6mua7l";

const CHARSET_REV = blk: {
    var table: [128]i8 = .{-1} ** 128;
    for (CHARSET, 0..) |c, i| {
        table[c] = @intCast(i);
    }
    break :blk table;
};

const GEN = [5]u32{ 0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3 };

fn polymod(values: []const u5) u32 {
    var chk: u32 = 1;
    for (values) |v| {
        const b = chk >> 25;
        chk = ((chk & 0x1ffffff) << 5) ^ @as(u32, v);
        inline for (0..5) |i| {
            if ((b >> @intCast(i)) & 1 != 0) {
                chk ^= GEN[i];
            }
        }
    }
    return chk;
}

/// Expand HRP for checksum computation: [high bits] ++ [0] ++ [low bits]
fn hrpExpand(hrp: []const u8, allocator: std.mem.Allocator) ![]u5 {
    var result = try allocator.alloc(u5, hrp.len * 2 + 1);
    for (hrp, 0..) |c, i| {
        result[i] = @truncate(c >> 5);
    }
    result[hrp.len] = 0;
    for (hrp, 0..) |c, i| {
        result[hrp.len + 1 + i] = @truncate(c & 0x1f);
    }
    return result;
}

// u5 is a Zig primitive type — no alias needed

fn verifyChecksum(hrp: []const u8, data: []const u5, allocator: std.mem.Allocator) !?Encoding {
    const hrp_exp = try hrpExpand(hrp, allocator);
    defer allocator.free(hrp_exp);

    var combined = try allocator.alloc(u5, hrp_exp.len + data.len);
    defer allocator.free(combined);
    @memcpy(combined[0..hrp_exp.len], hrp_exp);
    @memcpy(combined[hrp_exp.len..], data);

    const p = polymod(combined);
    if (p == BECH32_CONST) return .bech32;
    if (p == BECH32M_CONST) return .bech32m;
    return null;
}

fn createChecksum(hrp: []const u8, data: []const u5, encoding: Encoding, allocator: std.mem.Allocator) ![6]u5 {
    const hrp_exp = try hrpExpand(hrp, allocator);
    defer allocator.free(hrp_exp);

    var values = try allocator.alloc(u5, hrp_exp.len + data.len + 6);
    defer allocator.free(values);
    @memcpy(values[0..hrp_exp.len], hrp_exp);
    @memcpy(values[hrp_exp.len .. hrp_exp.len + data.len], data);
    @memset(values[hrp_exp.len + data.len ..], 0);

    const target: u32 = switch (encoding) {
        .bech32 => BECH32_CONST,
        .bech32m => BECH32M_CONST,
    };
    const p = polymod(values) ^ target;

    var result: [6]u5 = undefined;
    inline for (0..6) |i| {
        result[i] = @truncate((p >> @intCast(5 * (5 - i))) & 31);
    }
    return result;
}

// ─── Public API ──────────────────────────────────────────────────────────────

/// Encode raw 5-bit data with HRP into a bech32/bech32m string.
/// Low-level — most callers should use `encodeWitnessAddress` instead.
pub fn encode(hrp: []const u8, data: []const u5, encoding: Encoding, allocator: std.mem.Allocator) ![]u8 {
    const checksum = try createChecksum(hrp, data, encoding, allocator);
    const total_len = hrp.len + 1 + data.len + 6;
    var result = try allocator.alloc(u8, total_len);

    // HRP (lowercase)
    for (hrp, 0..) |c, i| {
        result[i] = if (c >= 'A' and c <= 'Z') c + 32 else c;
    }
    result[hrp.len] = '1'; // separator

    // Data characters
    for (data, 0..) |d, i| {
        result[hrp.len + 1 + i] = CHARSET[d];
    }
    // Checksum
    for (checksum, 0..) |d, i| {
        result[hrp.len + 1 + data.len + i] = CHARSET[d];
    }

    return result;
}

/// Decode a bech32/bech32m string → (hrp, 5-bit data, encoding).
/// Caller owns the returned slices.
pub fn decode(input: []const u8, allocator: std.mem.Allocator) !struct { hrp: []u8, data: []u5, encoding: Encoding } {
    // Find separator '1' — last occurrence
    var sep_pos: ?usize = null;
    for (0..input.len) |i| {
        if (input[input.len - 1 - i] == '1') {
            sep_pos = input.len - 1 - i;
            break;
        }
    }
    const pos = sep_pos orelse return error.InvalidBech32NoSeparator;

    if (pos < 1 or pos + 7 > input.len) return error.InvalidBech32TooShort;
    if (input.len > 90) return error.InvalidBech32TooLong;

    // Extract HRP (lowercase)
    var hrp = try allocator.alloc(u8, pos);
    for (input[0..pos], 0..) |c, i| {
        hrp[i] = if (c >= 'A' and c <= 'Z') c + 32 else c;
    }

    // Extract data part
    const data_part = input[pos + 1 ..];
    var data = try allocator.alloc(u5, data_part.len);
    for (data_part, 0..) |c, i| {
        const lower = if (c >= 'A' and c <= 'Z') c + 32 else c;
        if (lower >= 128) {
            allocator.free(hrp);
            allocator.free(data);
            return error.InvalidBech32Character;
        }
        const val = CHARSET_REV[lower];
        if (val < 0) {
            allocator.free(hrp);
            allocator.free(data);
            return error.InvalidBech32Character;
        }
        data[i] = @intCast(@as(u8, @bitCast(val)));
    }

    // Verify checksum
    const enc = try verifyChecksum(hrp, data, allocator) orelse {
        allocator.free(hrp);
        allocator.free(data);
        return error.InvalidBech32Checksum;
    };

    // Strip checksum from returned data
    const payload = try allocator.alloc(u5, data.len - 6);
    @memcpy(payload, data[0 .. data.len - 6]);
    allocator.free(data);

    return .{ .hrp = hrp, .data = payload, .encoding = enc };
}

// ─── Witness Address helpers (SegWit / Taproot) ──────────────────────────────

/// Convert between bit groups (e.g. 8→5 for encoding, 5→8 for decoding)
/// For 8→5: pass u8 data. For 5→8: pass u5 data.
pub fn convertBits8to5(data: []const u8, pad: bool, allocator: std.mem.Allocator) ![]u5 {
    return convertBitsGeneric(u8, data, 8, 5, pad, allocator);
}

pub fn convertBits5to8(data: []const u5, pad: bool, allocator: std.mem.Allocator) ![]u8 {
    return convertBitsGeneric(u5, data, 5, 8, pad, allocator);
}

fn convertBitsGeneric(comptime T: type, data: []const T, comptime from_bits: u5, comptime to_bits: u5, pad: bool, allocator: std.mem.Allocator) ![]std.meta.Int(.unsigned, to_bits) {
    const OutT = std.meta.Int(.unsigned, to_bits);
    var acc: u32 = 0;
    var bits: u6 = 0;
    const maxv: u32 = (@as(u32, 1) << to_bits) - 1;

    // Max output size
    const max_out = (data.len * @as(usize, from_bits) + @as(usize, to_bits) - 1) / @as(usize, to_bits);
    var result = try allocator.alloc(OutT, max_out + 1);
    var out_idx: usize = 0;

    for (data) |val| {
        acc = (acc << from_bits) | @as(u32, val);
        bits += from_bits;
        while (bits >= to_bits) {
            bits -= to_bits;
            result[out_idx] = @truncate((acc >> @intCast(bits)) & maxv);
            out_idx += 1;
        }
    }
    if (pad) {
        if (bits > 0) {
            result[out_idx] = @truncate((acc << @intCast(@as(u6, to_bits) - bits)) & maxv);
            out_idx += 1;
        }
    } else {
        if (bits >= from_bits) {
            allocator.free(result);
            return error.InvalidBitConversion;
        }
        if (((acc << @intCast(@as(u6, to_bits) - bits)) & maxv) != 0) {
            allocator.free(result);
            return error.NonZeroPadding;
        }
    }

    // Shrink to actual size
    const final = try allocator.alloc(OutT, out_idx);
    @memcpy(final, result[0..out_idx]);
    allocator.free(result);
    return final;
}

/// Encode a witness address: hrp + witness_version + witness_program (hash bytes)
/// witness_version=0 → Bech32 (P2WPKH: 20 bytes, P2WSH: 32 bytes)
/// witness_version=1 → Bech32m (Taproot: 32 bytes)
pub fn encodeWitnessAddress(
    hrp: []const u8,
    witness_version: u5,
    witness_program: []const u8,
    allocator: std.mem.Allocator,
) ![]u8 {
    // Validate witness program length
    if (witness_program.len < 2 or witness_program.len > 40) return error.InvalidWitnessLength;
    if (witness_version == 0 and witness_program.len != 20 and witness_program.len != 32)
        return error.InvalidV0WitnessLength;

    // Convert 8-bit → 5-bit
    const converted = try convertBits8to5(witness_program, true, allocator);
    defer allocator.free(converted);

    // Prepend witness version
    var data = try allocator.alloc(u5, 1 + converted.len);
    defer allocator.free(data);
    data[0] = witness_version;
    @memcpy(data[1..], converted);

    // Bech32 for v0, Bech32m for v1+
    const encoding: Encoding = if (witness_version == 0) .bech32 else .bech32m;

    return encode(hrp, data, encoding, allocator);
}

pub const WitnessResult = struct { version: u5, program: []u8 };

/// Decode a witness address → (witness_version, witness_program bytes)
/// Validates checksum, witness version, and program length per BIP-141/BIP-350.
pub fn decodeWitnessAddress(
    expected_hrp: []const u8,
    addr: []const u8,
    allocator: std.mem.Allocator,
) !WitnessResult {
    const decoded = try decode(addr, allocator);
    defer allocator.free(decoded.hrp);
    defer allocator.free(decoded.data);

    // Verify HRP
    if (!std.mem.eql(u8, decoded.hrp, expected_hrp)) return error.InvalidHRP;

    if (decoded.data.len < 1) return error.EmptyWitnessData;

    const version = decoded.data[0];
    if (version > 16) return error.InvalidWitnessVersion;

    // Verify encoding matches version
    const expected_enc: Encoding = if (version == 0) .bech32 else .bech32m;
    if (decoded.encoding != expected_enc) return error.WrongBech32Variant;

    // Convert 5-bit → 8-bit
    const program = try convertBits5to8(decoded.data[1..], false, allocator);

    // Validate program length
    if (program.len < 2 or program.len > 40) {
        allocator.free(program);
        return error.InvalidWitnessLength;
    }
    if (version == 0 and program.len != 20 and program.len != 32) {
        allocator.free(program);
        return error.InvalidV0WitnessLength;
    }

    return .{ .version = version, .program = program };
}

// ─── OmniBus convenience ────────────────────────────────────────────────────

pub const OB_HRP = "ob";

/// Generate an OmniBus P2WPKH address: ob1q... (witness v0, 20-byte hash160)
pub fn encodeOBAddress(hash160: [20]u8, allocator: std.mem.Allocator) ![]u8 {
    return encodeWitnessAddress(OB_HRP, 0, &hash160, allocator);
}

/// Generate an OmniBus Taproot address: ob1p... (witness v1, 32-byte x-only pubkey)
pub fn encodeOBTaprootAddress(pubkey_x: [32]u8, allocator: std.mem.Allocator) ![]u8 {
    return encodeWitnessAddress(OB_HRP, 1, &pubkey_x, allocator);
}

/// Validate and decode an OmniBus address (any witness version)
pub fn decodeOBAddress(addr: []const u8, allocator: std.mem.Allocator) !WitnessResult {
    return decodeWitnessAddress(OB_HRP, addr, allocator);
}

/// Quick validation — returns true if addr is a valid ob1... address
pub fn isValidOBAddress(addr: []const u8, allocator: std.mem.Allocator) bool {
    const decoded = decodeOBAddress(addr, allocator) catch return false;
    allocator.free(decoded.program);
    return true;
}

// ─── Tests ───────────────────────────────────────────────────────────────────

test "Bech32 — encode and decode roundtrip (witness v0, P2WPKH)" {
    const allocator = std.testing.allocator;

    // Hash160 known value (20 bytes)
    const hash160 = [20]u8{ 0x75, 0x1e, 0x76, 0xe8, 0x19, 0x91, 0x96, 0xd4, 0x54, 0x94, 0x1c, 0x45, 0xd1, 0xb3, 0xa3, 0x23, 0xf1, 0x43, 0x3b, 0xd6 };

    const addr = try encodeOBAddress(hash160, allocator);
    defer allocator.free(addr);

    // Must start with ob1q
    try std.testing.expect(std.mem.startsWith(u8, addr, "ob1q"));

    // Decode back
    const decoded = try decodeOBAddress(addr, allocator);
    defer allocator.free(decoded.program);

    try std.testing.expectEqual(@as(u5, 0), decoded.version);
    try std.testing.expectEqual(@as(usize, 20), decoded.program.len);
    try std.testing.expectEqualSlices(u8, &hash160, decoded.program);
}

test "Bech32m — encode and decode roundtrip (witness v1, Taproot)" {
    const allocator = std.testing.allocator;

    // 32-byte x-only pubkey
    var pubkey: [32]u8 = undefined;
    for (&pubkey, 0..) |*b, i| b.* = @truncate(i + 1);

    const addr = try encodeOBTaprootAddress(pubkey, allocator);
    defer allocator.free(addr);

    // Must start with ob1p (witness v1 = 'p' in bech32 charset)
    try std.testing.expect(std.mem.startsWith(u8, addr, "ob1p"));

    // Decode back
    const decoded = try decodeOBAddress(addr, allocator);
    defer allocator.free(decoded.program);

    try std.testing.expectEqual(@as(u5, 1), decoded.version);
    try std.testing.expectEqual(@as(usize, 32), decoded.program.len);
    try std.testing.expectEqualSlices(u8, &pubkey, decoded.program);
}

test "Bech32 — invalid checksum rejected" {
    const allocator = std.testing.allocator;

    // Generate valid address, then corrupt it
    const hash160 = [20]u8{ 0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff, 0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0xaa, 0xbb, 0xcc, 0xdd };
    const addr = try encodeOBAddress(hash160, allocator);
    defer allocator.free(addr);

    // Corrupt last character
    var bad = try allocator.alloc(u8, addr.len);
    defer allocator.free(bad);
    @memcpy(bad, addr);
    bad[bad.len - 1] = if (addr[addr.len - 1] == 'q') 'p' else 'q';

    try std.testing.expectError(error.InvalidBech32Checksum, decodeOBAddress(bad, allocator));
}

test "Bech32 — wrong encoding for witness version rejected" {
    const allocator = std.testing.allocator;

    // Encode v1 address with Bech32 (should be Bech32m) — manual encode
    var pubkey: [32]u8 = undefined;
    @memset(&pubkey, 0x42);

    const converted = try convertBits8to5(&pubkey, true, allocator);
    defer allocator.free(converted);

    var data = try allocator.alloc(u5, 1 + converted.len);
    defer allocator.free(data);
    data[0] = 1; // witness v1
    @memcpy(data[1..], converted);

    // Force Bech32 encoding (wrong for v1)
    const wrong_addr = try encode(OB_HRP, data, .bech32, allocator);
    defer allocator.free(wrong_addr);

    // Decoding should fail with WrongBech32Variant
    try std.testing.expectError(error.WrongBech32Variant, decodeOBAddress(wrong_addr, allocator));
}

test "Bech32 — isValidOBAddress" {
    const allocator = std.testing.allocator;

    const hash160 = [20]u8{ 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10, 0x11, 0x12, 0x13, 0x14 };
    const addr = try encodeOBAddress(hash160, allocator);
    defer allocator.free(addr);

    try std.testing.expect(isValidOBAddress(addr, allocator));
    try std.testing.expect(!isValidOBAddress("ob1qinvalid", allocator));
    try std.testing.expect(!isValidOBAddress("bc1qnotours", allocator));
    try std.testing.expect(!isValidOBAddress("", allocator));
}

test "Bech32 — deterministic encoding" {
    const allocator = std.testing.allocator;

    const hash160 = [20]u8{ 0xde, 0xad, 0xbe, 0xef, 0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff };

    const addr1 = try encodeOBAddress(hash160, allocator);
    defer allocator.free(addr1);
    const addr2 = try encodeOBAddress(hash160, allocator);
    defer allocator.free(addr2);

    try std.testing.expectEqualStrings(addr1, addr2);
}

File: core-rust/src/crypto/bech32.rs
// Bech32 / Bech32m encoder + decoder. Port of core/bech32.zig (BIP-173 + BIP-350).
//
// HRP = "ob" for OmniBus.
//   witness v0 -> Bech32 (P2WPKH ob1q... — 20-byte hash160; ob1q... for P2WSH 32B)
//   witness v1+ -> Bech32m (Taproot ob1p...)
//
// Output is byte-identical to core/bech32.zig — both implement the same
// BIP-173/350 polymod, charset, and bit-conversion.

pub const OB_HRP: &str = "ob";

const BECH32_CONST: u32 = 1;
const BECH32M_CONST: u32 = 0x2bc830a3;
const CHARSET: &[u8; 32] = b"qpzry9x8gf2tvdw0s3jn54khce6mua7l";
const GEN: [u32; 5] = [0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3];

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Encoding {
    Bech32,
    Bech32m,
}

#[derive(Debug, thiserror::Error)]
pub enum Bech32Error {
    #[error("no separator '1' in bech32 string")] NoSeparator,
    #[error("bech32 string too short")] TooShort,
    #[error("bech32 string too long (>90 chars)")] TooLong,
    #[error("invalid bech32 character")] InvalidChar,
    #[error("invalid bech32 checksum")] InvalidChecksum,
    #[error("wrong bech32 variant for witness version")] WrongVariant,
    #[error("invalid HRP (expected '{0}')")] InvalidHrp(String),
    #[error("invalid witness version")] InvalidWitnessVersion,
    #[error("invalid witness program length")] InvalidWitnessLength,
    #[error("invalid v0 witness program length (must be 20 or 32)")] InvalidV0WitnessLength,
    #[error("non-zero padding in bit conversion")] NonZeroPadding,
    #[error("invalid bit conversion")] InvalidBitConversion,
}

fn polymod(values: &[u8]) -> u32 {
    let mut chk: u32 = 1;
    for &v in values {
        let b = chk >> 25;
        chk = ((chk & 0x1ffffff) << 5) ^ (v as u32);
        for i in 0..5 {
            if ((b >> i) & 1) != 0 {
                chk ^= GEN[i];
            }
        }
    }
    chk
}

fn hrp_expand(hrp: &[u8]) -> Vec<u8> {
    let mut out = Vec::with_capacity(hrp.len() * 2 + 1);
    for &c in hrp { out.push(c >> 5); }
    out.push(0);
    for &c in hrp { out.push(c & 0x1f); }
    out
}

fn create_checksum(hrp: &[u8], data: &[u8], encoding: Encoding) -> [u8; 6] {
    let mut values = hrp_expand(hrp);
    values.extend_from_slice(data);
    values.extend_from_slice(&[0u8; 6]);
    let target = match encoding {
        Encoding::Bech32 => BECH32_CONST,
        Encoding::Bech32m => BECH32M_CONST,
    };
    let p = polymod(&values) ^ target;
    let mut out = [0u8; 6];
    for i in 0..6 {
        out[i] = ((p >> (5 * (5 - i))) & 31) as u8;
    }
    out
}

fn verify_checksum(hrp: &[u8], data: &[u8]) -> Option<Encoding> {
    let mut combined = hrp_expand(hrp);
    combined.extend_from_slice(data);
    let p = polymod(&combined);
    if p == BECH32_CONST { Some(Encoding::Bech32) }
    else if p == BECH32M_CONST { Some(Encoding::Bech32m) }
    else { None }
}

/// Encode 5-bit data + HRP to a bech32/bech32m string. `data` must contain
/// values in [0, 31].
pub fn encode(hrp: &str, data: &[u8], encoding: Encoding) -> String {
    let checksum = create_checksum(hrp.as_bytes(), data, encoding);
    let mut s = String::with_capacity(hrp.len() + 1 + data.len() + 6);
    for c in hrp.chars() { s.push(c.to_ascii_lowercase()); }
    s.push('1');
    for &d in data { s.push(CHARSET[d as usize] as char); }
    for &d in &checksum { s.push(CHARSET[d as usize] as char); }
    s
}

/// Decode a bech32/bech32m string. Returns (hrp, 5-bit data including checksum-stripped, encoding).
pub fn decode(input: &str) -> Result<(String, Vec<u8>, Encoding), Bech32Error> {
    if input.len() > 90 { return Err(Bech32Error::TooLong); }
    let bytes = input.as_bytes();
    // last '1'
    let pos = bytes.iter().rposition(|&c| c == b'1').ok_or(Bech32Error::NoSeparator)?;
    if pos < 1 || pos + 7 > bytes.len() { return Err(Bech32Error::TooShort); }

    let mut hrp = String::with_capacity(pos);
    for &c in &bytes[..pos] {
        hrp.push((c as char).to_ascii_lowercase());
    }

    let data_part = &bytes[pos + 1..];
    let mut data = Vec::with_capacity(data_part.len());
    for &c in data_part {
        let lower = if (b'A'..=b'Z').contains(&c) { c + 32 } else { c };
        if lower >= 128 { return Err(Bech32Error::InvalidChar); }
        let idx = CHARSET.iter().position(|&x| x == lower).ok_or(Bech32Error::InvalidChar)? as u8;
        data.push(idx);
    }

    let enc = verify_checksum(hrp.as_bytes(), &data).ok_or(Bech32Error::InvalidChecksum)?;
    data.truncate(data.len() - 6);
    Ok((hrp, data, enc))
}

/// 8-bit -> 5-bit conversion with optional padding (for encoding witness program).
pub fn convert_bits(data: &[u8], from_bits: u32, to_bits: u32, pad: bool) -> Result<Vec<u8>, Bech32Error> {
    let mut acc: u32 = 0;
    let mut bits: u32 = 0;
    let maxv: u32 = (1u32 << to_bits) - 1;
    let mut out = Vec::with_capacity((data.len() * from_bits as usize + to_bits as usize - 1) / to_bits as usize + 1);
    for &val in data {
        if (val as u32) >> from_bits != 0 { return Err(Bech32Error::InvalidBitConversion); }
        acc = (acc << from_bits) | (val as u32);
        bits += from_bits;
        while bits >= to_bits {
            bits -= to_bits;
            out.push(((acc >> bits) & maxv) as u8);
        }
    }
    if pad {
        if bits > 0 {
            out.push(((acc << (to_bits - bits)) & maxv) as u8);
        }
    } else if bits >= from_bits || ((acc << (to_bits - bits)) & maxv) != 0 {
        return Err(Bech32Error::NonZeroPadding);
    }
    Ok(out)
}

/// Encode a witness address (hrp + witness_version + 8-bit program bytes).
/// v0 -> Bech32, v1+ -> Bech32m.
pub fn encode_witness_address(hrp: &str, witness_version: u8, witness_program: &[u8]) -> Result<String, Bech32Error> {
    if witness_program.len() < 2 || witness_program.len() > 40 {
        return Err(Bech32Error::InvalidWitnessLength);
    }
    if witness_version == 0 && witness_program.len() != 20 && witness_program.len() != 32 {
        return Err(Bech32Error::InvalidV0WitnessLength);
    }
    if witness_version > 16 { return Err(Bech32Error::InvalidWitnessVersion); }

    let converted = convert_bits(witness_program, 8, 5, true)?;
    let mut data = Vec::with_capacity(1 + converted.len());
    data.push(witness_version);
    data.extend_from_slice(&converted);

    let encoding = if witness_version == 0 { Encoding::Bech32 } else { Encoding::Bech32m };
    Ok(encode(hrp, &data, encoding))
}

#[derive(Debug, Clone)]
pub struct WitnessResult {
    pub version: u8,
    pub program: Vec<u8>,
}

pub fn decode_witness_address(expected_hrp: &str, addr: &str) -> Result<WitnessResult, Bech32Error> {
    let (hrp, data, enc) = decode(addr)?;
    if hrp != expected_hrp { return Err(Bech32Error::InvalidHrp(expected_hrp.to_string())); }
    if data.is_empty() { return Err(Bech32Error::InvalidWitnessLength); }
    let version = data[0];
    if version > 16 { return Err(Bech32Error::InvalidWitnessVersion); }
    let expected_enc = if version == 0 { Encoding::Bech32 } else { Encoding::Bech32m };
    if enc != expected_enc { return Err(Bech32Error::WrongVariant); }

    let program = convert_bits(&data[1..], 5, 8, false)?;
    if program.len() < 2 || program.len() > 40 { return Err(Bech32Error::InvalidWitnessLength); }
    if version == 0 && program.len() != 20 && program.len() != 32 {
        return Err(Bech32Error::InvalidV0WitnessLength);
    }
    Ok(WitnessResult { version, program })
}

/// Encode an OmniBus P2WPKH address: ob1q... (HRP="ob", v0, 20-byte hash160).
pub fn encode_ob_address(hash160: &[u8; 20]) -> String {
    // unwrap: 20-byte program, v0 -> always valid
    encode_witness_address(OB_HRP, 0, hash160).expect("ob P2WPKH encode")
}

/// Encode an OmniBus Taproot address: ob1p... (HRP="ob", v1, 32-byte x-only pubkey).
pub fn encode_ob_taproot_address(pubkey_x: &[u8; 32]) -> String {
    encode_witness_address(OB_HRP, 1, pubkey_x).expect("ob taproot encode")
}

pub fn decode_ob_address(addr: &str) -> Result<WitnessResult, Bech32Error> {
    decode_witness_address(OB_HRP, addr)
}

#[cfg(test)]
mod tests {
    use super::*;

    // Vector mirrors core/bech32.zig "Bech32 — encode and decode roundtrip (witness v0, P2WPKH)"
    // hash160 = 751e76e819919...3bd6 -> "ob1q..." starts.
    #[test]
    fn ob1q_roundtrip() {
        let hash160: [u8; 20] = [
            0x75, 0x1e, 0x76, 0xe8, 0x19, 0x91, 0x96, 0xd4, 0x54, 0x94,
            0x1c, 0x45, 0xd1, 0xb3, 0xa3, 0x23, 0xf1, 0x43, 0x3b, 0xd6,
        ];
        let addr = encode_ob_address(&hash160);
        assert!(addr.starts_with("ob1q"), "got {addr}");
        let decoded = decode_ob_address(&addr).unwrap();
        assert_eq!(decoded.version, 0);
        assert_eq!(decoded.program, hash160.to_vec());
    }

    #[test]
    fn taproot_roundtrip() {
        let mut pk = [0u8; 32];
        for (i, b) in pk.iter_mut().enumerate() { *b = (i as u8).wrapping_add(1); }
        let addr = encode_ob_taproot_address(&pk);
        assert!(addr.starts_with("ob1p"), "got {addr}");
        let decoded = decode_ob_address(&addr).unwrap();
        assert_eq!(decoded.version, 1);
        assert_eq!(decoded.program, pk.to_vec());
    }

    #[test]
    fn rejects_corrupt_checksum() {
        let h: [u8; 20] = [0xaa; 20];
        let addr = encode_ob_address(&h);
        let mut bad: Vec<u8> = addr.into_bytes();
        let last = bad.len() - 1;
        bad[last] = if bad[last] == b'q' { b'p' } else { b'q' };
        let s = String::from_utf8(bad).unwrap();
        assert!(decode_ob_address(&s).is_err());
    }
}
